modules/currencies/src/lib.rs

```diff
@@ -369,46 +388,42 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
 			return Ok(());
 		}
 		match currency_id {
-			CurrencyId::Erc20(_) => return Err(Error::<T>::Erc20InvalidOperation.into()),
-			id if id == T::GetNativeCurrencyId::get() => T::NativeCurrency::deposit(who, amount)?,
-			_ => T::MultiCurrency::deposit(currency_id, who, amount)?,
+			CurrencyId::Erc20(_) => Err(Error::<T>::Erc20InvalidOperation.into()),
+			id if id == T::GetNativeCurrencyId::get() => T::NativeCurrency::deposit(who, amount),
+			_ => T::MultiCurrency::deposit(currency_id, who, amount),
 		}
-		Self::deposit_event(Event::Deposited {
-			currency_id,
-			who: who.clone(),
-			amount,
-		});
-		Ok(())
 	}
 
 	fn withdraw(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
 		if amount.is_zero() {
 			return Ok(());
 		}
+
 		match currency_id {
-			CurrencyId::Erc20(_) => return Err(Error::<T>::Erc20InvalidOperation.into()),
-			id if id == T::GetNativeCurrencyId::get() => T::NativeCurrency::withdraw(who, amount)?,
-			_ => T::MultiCurrency::withdraw(currency_id, who, amount)?,
+			CurrencyId::Erc20(_) => Err(Error::<T>::Erc20InvalidOperation.into()),
+			id if id == T::GetNativeCurrencyId::get() => T::NativeCurrency::withdraw(who, amount),
+			_ => T::MultiCurrency::withdraw(currency_id, who, amount),
 		}
-		Self::deposit_event(Event::Withdrawn {
-			currency_id,
-			who: who.clone(),
-			amount,
-		});
-		Ok(())
 	}
 
 	fn can_slash(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> bool {
 		match currency_id {
-			CurrencyId::Erc20(_) => false,
+			CurrencyId::Erc20(_) => amount.is_zero(),
 			id if id == T::GetNativeCurrencyId::get() => T::NativeCurrency::can_slash(who, amount),
 			_ => T::MultiCurrency::can_slash(currency_id, who, amount),
 		}
 	}
 
 	fn slash(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> Self::Balance {
 		match currency_id {
-			CurrencyId::Erc20(_) => Default::default(),
+			CurrencyId::Erc20(_) => {
+				Self::deposit_event(Event::Slashed {
```

Comment:
this should be `Erc20InvalidOperation`?

Reply:
The function signature require return of "Balance" - so error cannot be returned